src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginSubHandlers/PreloginPacketHandler.cs

```diff
@@ -0,0 +1,412 @@
+                        {
+                            SqlClientEventSource.Log.TryTraceEvent("<sc.{0}|ERR> {1}, " +
+                                "Server sent an unexpected value for FedAuthRequired PreLogin Option. Value was {2}.", nameof(ReadPreLoginresponse), ObjectID, (int)payload[payloadOffset]);
+                            throw SQL.ParsingErrorValue(ParsingErrorState.FedAuthRequiredPreLoginResponseInvalidValue, (int)payload[payloadOffset]);
+                        }
+
+                        // We must NOT use the response for the FEDAUTHREQUIRED PreLogin option, if the connection string option
+                        // was not using the new Authentication keyword or in other words, if Authentication=NotSpecified
+                        // Or AccessToken is not null, mean token based authentication is used.
+                        if ((context.ConnectionContext.ConnectionString != null
+                            && context.ConnectionContext.ConnectionString.Authentication != SqlAuthenticationMethod.NotSpecified)
+                            || context.ConnectionContext.AccessTokenInBytes != null || context.ConnectionContext.AccessTokenCallback != null)
+                        {
+                            context.ConnectionContext.FedAuthRequired = payload[payloadOffset] == 0x01;
+                        }
+                        break;
+
+                    default:
+                        Debug.Fail("UNKNOWN option in ConsumePreLoginHandshake, option:" + option);
+                        // DO NOTHING FOR THESE UNKNOWN OPTIONS
+                        offset += PAYLOAD_OFFSET_AND_LENGTH_SIZE_IN_BYTES;
+                        break;
+                }
+
+                if (offset < payload.Length)
+                {
+                    option = payload[offset++];
+                }
+                else
+                {
+                    break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Constructs the Prelogin packet and sends it to the server.
+        /// </summary>
+        /// <param name="context">The Prelogin handler context</param>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private static async Task CreatePreLoginAndSend(PreLoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            Debug.Assert(context.ConnectionContext.TdsStream != null, "A Tds Stream is expected");
+
+            TdsStream tdsStream = context.ConnectionContext.TdsStream;
+            tdsStream.PacketHeaderType = TdsStreamPacketType.PreLogin;
+
+            // Initialize option offset into payload buffer
+            // 5 bytes for each option (1 byte length, 2 byte offset, 2 byte payload length)
+            int offset = (int)PreLoginOptions.NUMOPT * 5 + 1;
+
+            byte[] payload = new byte[(int)PreLoginOptions.NUMOPT * 5 + TdsEnums.MAX_PRELOGIN_PAYLOAD_LENGTH];
```

Comment:
The payload is of fixed length - a lot smaller than this. It'd be better to calculate this to minimise the allocation, making sure that it includes both the data and the header section (as I think it does now.)

As before: in the sync case, this can be a stackalloc'd Span. Since this is a fixed-length payload, I'd also rent it from an `ArrayPool` for the async case.